### PR TITLE
Correlates client-side and server segment anonymous_id's

### LINKED ISSUE #253: Analytics: Correlate Client- and Server-side Tracking
Most of #179 has shipped now, but the conversion funnels in GA don't look right. My hypothesis is that `anonymousId` needs to match throughout a session. I'd suggest trying to set Rails `session.id` as the `anonymousId` on both client- and server-side requests.`

--------------------------

### Client-side `/login` page:

```json
{
  "anonymousId": "3d22d5be-eddf-46a5-aed8-bf482420d814",
  "context": {
    "page": {
      "path": "/login",
      "referrer": "https://huboard.com/",
      "search": "",
      "title": "Huboard",
      "url": "https://huboard.com/login"
    },
    "userAgent": "Mozilla/5.0 (Windows NT 6.3; WOW64; rv:45.0) Gecko/20100101 Firefox/45.0",
    "library": {
      "name": "analytics.js",
      "version": "2.11.1"
    },
    "ip": "50.83.133.203"
  },
  "integrations": {},
  "messageId": "ajs-0014fb9e8964a9bc622ffa16a94564ba",
  "properties": {
    "path": "/login",
    "referrer": "https://huboard.com/",
    "search": "",
    "title": "Huboard",
    "url": "https://huboard.com/login"
  },
  "receivedAt": "2016-03-25T19:40:25.809Z",
  "sentAt": "2016-03-25T19:40:23.746Z",
  "timestamp": "2016-03-25T19:40:25.808Z",
  "type": "page",
  "userId": "133987",
  "originalTimestamp": "2016-03-25T19:40:23.745Z"
}
```

### Server-side `/login/public` page:

```json
{
  "userId": "Anonymous",
  "anonymousId": null,
  "name": "/login/public",
  "category": null,
  "properties": {
    "url": "/login/public"
  },
  "integrations": {},
  "context": {
    "library": {
      "name": "analytics-ruby",
      "version": "2.0.13"
    }
  },
  "timestamp": "2016-03-25T19:40:27.419Z",
  "type": "page",
  "messageId": "58990ac4-8972-476c-9edc-f8fa7a28effc",
  "writeKey": "trrvdc58v4",
  "sentAt": "2016-03-25T19:40:26.964Z",
  "receivedAt": "2016-03-25T19:40:27.422Z",
  "originalTimestamp": "2016-03-25T19:40:26.961+00:00"
}
```

### Server-side identify:

```json
{
  "userId": "133987",
  "anonymousId": null,
  "integrations": {},
  "context": {
    "library": {
      "name": "analytics-ruby",
      "version": "2.0.13"
    }
  },
  "traits": {
    "login": "dahlbyk",
    "id": 133987,
    "avatar_url": "https://avatars.githubusercontent.com/u/133987?v=3",
    "gravatar_id": "",
    "url": "https://api.github.com/users/dahlbyk",
    "html_url": "https://github.com/dahlbyk",
    "followers_url": "https://api.github.com/users/dahlbyk/followers",
    "following_url": "https://api.github.com/users/dahlbyk/following{/other_user}",
    "gists_url": "https://api.github.com/users/dahlbyk/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/dahlbyk/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/dahlbyk/subscriptions",
    "organizations_url": "https://api.github.com/users/dahlbyk/orgs",
    "repos_url": "https://api.github.com/users/dahlbyk/repos",
    "events_url": "https://api.github.com/users/dahlbyk/events{/privacy}",
    "received_events_url": "https://api.github.com/users/dahlbyk/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Keith Dahlby",
    "company": null,
    "blog": "http://keith.lostechies.com/",
    "location": "Cedar Rapids, IA",
    "email": "dahlbyk@gmail.com",
    "hireable": null,
    "bio": null,
    "public_repos": 57,
    "public_gists": 19,
    "followers": 174,
    "following": 11,
    "created_at": "2009-10-02T02:13:10Z",
    "updated_at": "2016-03-10T20:12:11Z",
    "email_verified": true,
    "emails": [
      {
        "email": "dahlbyk@gmail.com",
        "primary": true,
        "verified": true
      },
      {
        "email": "dahlbyk@huboard.com",
        "primary": false,
        "verified": true
      }
    ],
    "action": "public_login"
  },
  "timestamp": "2016-03-25T19:40:28.243Z",
  "type": "identify",
  "messageId": "2fd808ab-d8be-479b-8b39-7876b9ba48ab",
  "writeKey": "trrvdc58v4",
  "sentAt": "2016-03-25T19:40:27.829Z",
  "receivedAt": "2016-03-25T19:40:28.244Z",
  "originalTimestamp": "2016-03-25T19:40:27.828+00:00"
}
```

### Server-side `/login/public/authorized` page:

```json
{
  "userId": "133987",
  "anonymousId": null,
  "name": "/login/public/authorized",
  "category": null,
  "properties": {
    "url": "/login/public/authorized"
  },
  "integrations": {},
  "context": {
    "library": {
      "name": "analytics-ruby",
      "version": "2.0.13"
    }
  },
  "timestamp": "2016-03-25T19:40:28.235Z",
  "type": "page",
  "messageId": "299aab1b-209a-4576-8447-eda4d1a1f587",
  "writeKey": "trrvdc58v4",
  "sentAt": "2016-03-25T19:40:28.255Z",
  "receivedAt": "2016-03-25T19:40:28.649Z",
  "originalTimestamp": "2016-03-25T19:40:27.841+00:00"
}
```

### Client-side identify:

```json
{
  "anonymousId": "3d22d5be-eddf-46a5-aed8-bf482420d814",
  "context": {
    "page": {
      "path": "/",
      "referrer": "https://huboard.com/login",
      "search": "",
      "title": "HuBoard",
      "url": "https://huboard.com/"
    },
    "userAgent": "Mozilla/5.0 (Windows NT 6.3; WOW64; rv:45.0) Gecko/20100101 Firefox/45.0",
    "library": {
      "name": "analytics.js",
      "version": "2.11.1"
    },
    "ip": "50.83.133.203"
  },
  "integrations": {},
  "messageId": "ajs-061c3735acc916e8e293e07acdf25c5d",
  "receivedAt": "2016-03-25T19:40:30.007Z",
  "sentAt": "2016-03-25T19:40:27.825Z",
  "timestamp": "2016-03-25T19:40:30.002Z",
  "traits": {
    "login": "dahlbyk",
    "id": 133987,
    "avatar_url": "https://avatars.githubusercontent.com/u/133987?v=3",
    "gravatar_id": "",
    "site_admin": false,
    "name": "Keith Dahlby",
    "company": "HuBoard",
    "email": "dahlbyk@gmail.com"
  },
  "type": "identify",
  "userId": "133987",
  "originalTimestamp": "2016-03-25T19:40:27.820Z"
}
```

### Client-side authenticated `/` page:

```json
{
  "anonymousId": "3d22d5be-eddf-46a5-aed8-bf482420d814",
  "category": null,
  "context": {
    "page": {
      "path": "/",
      "referrer": "https://huboard.com/login",
      "search": "",
      "title": "HuBoard",
      "url": "https://huboard.com/"
    },
    "userAgent": "Mozilla/5.0 (Windows NT 6.3; WOW64; rv:45.0) Gecko/20100101 Firefox/45.0",
    "library": {
      "name": "analytics.js",
      "version": "2.11.1"
    },
    "ip": "50.83.133.203"
  },
  "integrations": {},
  "messageId": "ajs-3e53dd2a6e87a027b99e91655f3881c9",
  "name": null,
  "properties": {
    "userId": "",
    "path": "/",
    "referrer": "https://huboard.com/login",
    "search": "",
    "title": "HuBoard",
    "url": "https://huboard.com/"
  },
  "receivedAt": "2016-03-25T19:40:30.116Z",
  "sentAt": "2016-03-25T19:40:27.838Z",
  "timestamp": "2016-03-25T19:40:30.115Z",
  "type": "page",
  "userId": "133987",
  "originalTimestamp": "2016-03-25T19:40:27.837Z"
}
```

### PR DESCRIPTION
**Pull request by discorick**
Closes #253 #162

### CODE REVIEW

**Comment by discorick** [comment]
debugger_-_segment

My only concern is the middle step in the pipeline, the `page()` method doesn't seem to want to overwrite the `anonymous_id`, I am not sure how the GA integration will handle collapsing that and the properties hash.

Starting from the bottom highlighted payload:

```json
{
  "anonymousId": "f9275b47-dbb6-4bba-b3f8-db5f871dbefc",
  "category": null,
  "context": {
    "page": {
      "path": "/login",
      "referrer": "http://huboard.rails/",
      "search": "",
      "title": "Huboard",
      "url": "http://huboard.rails/login"
    },
    "userAgent": "blahblahblah",
    "library": {
      "name": "analytics.js",
      "version": "2.11.1"
    },
    "ip": "68.146.230.217"
  },
  "integrations": {},
  "messageId": "ajs-bf209c1cc0373d8215d9fcc459a4fb23",
  "name": null,
  "properties": {
    "anonymousId": "5bc579b368d3ec787c1355a561d533ac",
    "path": "/login",
    "referrer": "http://huboard.rails/",
    "search": "",
    "title": "Huboard",
    "url": "http://huboard.rails/login"
  },
  "receivedAt": "2016-03-28T18:33:58.514Z",
  "sentAt": "2016-03-28T18:33:58.487Z",
  "timestamp": "2016-03-28T18:33:58.513Z",
  "type": "page",
  "userId": "1130665",
  "originalTimestamp": "2016-03-28T18:33:58.486Z"
}
```

```json
{
  "userId": "Anonymous",
  "anonymousId": "5bc579b368d3ec787c1355a561d533ac",
  "name": "/login/private",
  "category": null,
  "properties": {
    "url": "/login/private"
  },
  "integrations": {},
  "context": {
    "library": {
      "name": "analytics-ruby",
      "version": "2.0.13"
    }
  },
  "timestamp": "2016-03-28T18:33:59.818Z",
  "type": "page",
  "messageId": "b30b5b83-5820-4012-8d75-c54eb9f7316d",
  "writeKey": "redacted",
  "sentAt": "2016-03-28T18:34:34.448Z",
  "receivedAt": "2016-03-28T18:34:34.851Z",
  "originalTimestamp": "2016-03-28T11:33:59.415-07:00"
}
```
```json
{
  "userId": "1130665",
  "anonymousId": "9a3d0d56f08f65401416196a5dff89cc",
  "name": "/login/private/authorized",
  "category": null,
  "properties": {
    "url": "/login/private/authorized"
  },
  "integrations": {},
  "context": {
    "library": {
      "name": "analytics-ruby",
      "version": "2.0.13"
    }
  },
  "timestamp": "2016-03-28T18:34:07.068Z",
  "type": "page",
  "messageId": "0b5af0e8-1096-4b48-8fed-5e2aa8e3cf83",
  "writeKey": "redacted",
  "sentAt": "2016-03-28T18:34:34.723Z",
  "receivedAt": "2016-03-28T18:34:35.146Z",
  "originalTimestamp": "2016-03-28T11:34:06.645-07:00"
}
```

File: app/views/shared/_analytics.html.erb [hunk 1]
```diff
@@ -6,7 +6,8 @@
     window.analytics.identify('<%= current_user.id %>', {})
     window.analytics.page({ userId: '<% current_user.id %>' });
   <% else %>
-    window.analytics.page();
+      window.analytics.identify({ anonymousId: '<%= session.id %>' });
+      window.analytics.page({ anonymousId: '<%= session.id %>' });
```
Comment (dahlbyk) [inline app/views/shared/_analytics.html.erb]: Should we also set `anonymousId` when `logged_in?`?

Comment (discorick) [inline app/views/shared/_analytics.html.erb]: Yeah I think so.

File: vendor/engines/saas/app/jobs/analytics/page_job.rb [hunk 2]
```diff
@@ -6,6 +6,7 @@ def payload(params)
       user = params['current_user'] ? params['current_user']['id'] : "Anonymous"
       {
         user_id: user,
+        anonymous_id: params['session_id'],
```
Comment (dahlbyk) [inline vendor/engines/saas/app/jobs/analytics/page_job.rb]: Should this be `anonymousId` rather than `anonymous_id`? Same question for `IdentifyUserJob`.

Comment (dahlbyk) [inline vendor/engines/saas/app/jobs/analytics/page_job.rb]: Nevermind, checked docs.

**Comment by discorick** [comment]
So, oddly enough Warden sees fit to instruct rack to reset the session id after every login. I have tried getting around this using callbacks:

```ruby 
Warden::Manager.after_set_user do |user,auth,opts|
  auth.env["rack.session.options"][:renew] = false
end
```

That didn't work, switching `after_set_user` with `on_request` callback didn't help either, both happen after the rack session id has been renewed.

File: config/initializers/warden.rb [hunk 3]
```diff
@@ -75,4 +75,26 @@ def call(env)
...
+
+    # Get the default options from the master configuration for the given scope
+    opts = (@config[:scope_defaults][scope] || {}).merge(opts)
+    opts[:event] ||= :set_user
+    @users[scope] = user
+
+    if opts[:store] != false && opts[:event] != :fetch
+      options = env[ENV_SESSION_OPTIONS]
+      session_serializer.store(user, scope)
+    end
```
Comment (dahlbyk) [inline config/initializers/warden.rb]: For our future reference, `set_user` is copy/pasta from https://github.com/hassox/warden/blob/cb09b959bcac82a087ff0f2938273e6924d845d8/lib/warden/proxy.rb#L172-L176 except:

```diff
   if opts[:store] != false && opts[:event] != :fetch
     options = env[ENV_SESSION_OPTIONS]
-    options[:renew] = true if options
     session_serializer.store(user, scope)
   end
```

I know we're considering a switch to OmniAuth (#157), but I wonder if we should submit a patch to Warden along the lines of:

```diff

```diff
   if opts[:store] != false && opts[:event] != :fetch
     options = env[ENV_SESSION_OPTIONS]
-    options[:renew] = true if options
+    options[:renew] = true if options && !opts[:renew]
     session_serializer.store(user, scope)
   end
```

**Comment by dahlbyk** [comment]
So the session reset seems to be attack mitigation technique, which we probably don't want to lose just in the name of analytics.

- https://github.com/monterail/guidelines/issues/166
- https://github.com/plataformatec/devise/issues/3706
- https://github.com/hassox/warden/commit/d595ae19793080178af8d27d61eed6b54f620150

Two options:
- Read Segment's `ajs_anonymous_id` cookie from each request and pass it down through.
- Add something server-side to inject our own correlation ID into `session` when one is created, and ensure that value is preserved through login.

The latter seems more durable in the presence of ad blockers.

**Comment by discorick** [comment]
@dahlbyk , I opted for option number 2 - your right, session reset was designed to prevent session fixation attacks.

**Comment by dahlbyk** [comment]
Also, while we're here: I noticed that we're still including the previous `userId` in payloads even after Logout (between pages `/repositories/private/dahlbyk` and `/` at the top):

firefox_2016-03-31_09-24-01

Speaking of which, we should probably track `/logout` server-side before actually logging out.

**Comment by dahlbyk** [comment]
> Try yet another approach to set anonymousId client-side

This seems to be working as expected. My only doubt now is that I _appear_ to have seen my `guid` change when clicking from my initial dashboard (via `/login/public`) to the Private tab. Not sure if there's something weird happening there with the bump in auth scope, or if I'm just seeing things.

**Comment by discorick** [comment]
>  My only doubt now is that I appear to have seen my guid change when clicking from my initial dashboard (via /login/public) to the Private tab

You are correct, when we upgrade auth scope this is happening: 

`request.env['warden'].logout if github_authenticated? :default` effectively clearing the session entirely (see `app/controllers/dashboard_controller.rb` LN 47)

The question is should we fix this now, or as part of our improved login flow?

**Comment by dahlbyk** [comment]
> The question is should we fix this now, or as part of our improved login flow?

What happens if we just delete that line ([and](https://github.com/huboard/huboard-web/blob/87123f849beef159ef35502266506d8aa27caa7a/app/controllers/login_controller.rb#L8) [these](https://github.com/huboard/huboard-web/blob/87123f849beef159ef35502266506d8aa27caa7a/app/controllers/login_controller.rb#L15))? AFAICT, the logout only seems to be a problem from an Analytics session cohesion standpoint.

**Comment by discorick** [comment]
After removing the explicit logouts on scope change everything seems to still function as it should, and our anon `guid` remains consistent across the board. 

@rauhryan , I remember way back when we moved to rails there was a reason for the logouts - it seems thats no longer the case (see 264d998 )?

**Comment by dahlbyk** [comment]
While we wait for @rauhryan to refresh his memory, can you check on resetting `userId` on Logout:

> I noticed that we're still including the previous `userId` in payloads even after Logout (between pages /repositories/private/dahlbyk and / at the top)

I would expect the session to be reset altogether (`anonymousId`, too).

**Comment by discorick** [comment]
It looks like Segment is caching the traits client-side, so even though the session is lost `analytics.js` can still tell who the user is (this includes all the keys in the traits object)

I don't see any inherent risk here as we are only storing publicly available GitHub data.

**Comment by dahlbyk** [comment]
> I don't see any inherent risk here as we are only storing publicly available GitHub data.

It's just a correctness thing. Analytics.js does provide [`reset()`](https://segment.com/docs/libraries/analytics.js/#reset-logout) that we could inject client-side. I'd rather ship this than worry too much about tracking after logout.

**Comment by dahlbyk** [comment]
> After removing the explicit logouts on scope change everything seems to still function as it should, and our anon `guid` remains consistent across the board.
> 
> @rauhryan , I remember way back when we moved to rails there was a reason for the logouts - it seems thats no longer the case (see 264d998 )?

I've kicked the tires quite a bit on this...not seeing anything unexpected through various login/logout cycles starting from public and private and then upgrading or downgrading. Merging.

**Comment by rauhryan** [comment]
@discorick @dahlbyk 

I remember the reason for adding the explicit logout was to force an OAuth change request to add the ability to "downgrade" HuBoard's access

But it sounds like GitHub has changed the default behavior of their OAuth flow

**Pull request merged**